drivers/volume/portworx/portworx.go

```diff
@@ -550,12 +623,104 @@ func (p *portworx) getSnapshotName(tags *map[string]string) string {
 	return "snapshot-" + (*tags)[snapshotter.CloudSnapshotCreatedForVolumeSnapshotUIDTag]
 }
 
+func (p *portworx) getUserContext(ctx context.Context, annotations map[string]string) (context.Context, error) {
+	if v, ok := annotations[auth_secrets.SecretNameKey]; ok {
+		token, err := p.authSecrets.GetToken(v, annotations[auth_secrets.SecretNamespaceKey])
+		if err != nil {
+			return nil, err
+		}
+		return grpcserver.AddMetadataToContext(ctx, "authorization", "bearer "+token), nil
+	}
+	return ctx, nil
+}
+
+func (p *portworx) getUserVolDriver(annotations map[string]string) (volume.VolumeDriver, error) {
+	if v, ok := annotations[auth_secrets.SecretNameKey]; ok {
+		token, err := p.authSecrets.GetToken(v, annotations[auth_secrets.SecretNamespaceKey])
+		if err != nil {
+			return nil, err
+		}
+		clnt, err := p.getRestClientWithAuth(token)
+		if err != nil {
+			return nil, err
+		}
+		return volumeclient.VolumeDriver(clnt), nil
+	}
+
+	clnt, err := p.getRestClient()
+	if err != nil {
+		return nil, err
+	}
+	return volumeclient.VolumeDriver(clnt), nil
+
+}
+
+func (p *portworx) getAdminVolDriver() (volume.VolumeDriver, error) {
+	if len(p.jwtSharedSecret) != 0 {
+		claims := &auth.Claims{
+			Issuer: "stork.openstorage.io",
+			Name:   "Stork",
+
+			// Unique id for stork
+			// this id must be unique across all accounts accessing the px system
+			Subject: "stork.openstorage.io." + p.id,
+
+			// Only allow certain calls
+			Roles: []string{"system.user"},
+
+			// Be in all groups to have access to all resources
+			Groups: []string{"*"},
+		}
+
+		// This never returns an error, but just in case, check the value
+		signature, err := auth.NewSignatureSharedSecret(p.jwtSharedSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set the token expiration
+		options := &auth.Options{
+			Expiration: time.Now().Add(time.Hour).Unix(),
+		}
+
+		token, err := auth.Token(claims, signature, options)
```

Comment:
Why generate this token for every call? We should store it and regenerate when it expires.

Reply:
This is a good point, but that would mean tracking tokens with time. If a token is expired generated a new one, and so on. I think we can do that, but that complexity I would like to move to another PR for later, if you don't mind. Currently generating a token each time is very fast.

Reply:
Generating a token is expensive if done very often, they involve crypto calls. Please file a bug to track that.
Example of how kubernetes deals with OIDC tokens to alleviate this: https://github.com/kubernetes/client-go/blob/master/plugin/pkg/client/auth/oidc/oidc.go#L332

Reply:
OIDC is different because it is an RPC connection to get the public auth key. We only generate it. But I'm not against it. I can change it to do this, I just suggested we do this after.

Reply:
Filed #325